doc/ctapi.3
.\" This -*- nroff -*- file has been generated from
.\" DocBook SGML with docbook-to-man on Debian GNU/Linux.
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
\\&.
.fl			\" force out current output buffer
\\!%PB
\\!/showpage{}def
...\" the following is from Ken Flowers -- it prevents dictionary overflows
\\!/tempdict 200 dict def tempdict begin
.fl			\" prolog
.sy cat \\$1\" bring in postscript file
...\" the following line matches the tempdict above
\\!end % tempdict %
\\!PE
\\!.
.sp \\$2u	\" move below the image
..
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
'\" t 
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n  
.TH "CT-API" "3" 
.SH "NAME" 
\fBCT_init, CT_data, CT_close\fR  \(em 	Functions to handle communication with integrated circuit cards and  
cardterminals 
 
.SH "SYNOPSIS" 
.PP 
.nf 
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n 
#include <ctapi.h> 
.sp 1 
\fBchar \fBCT_init\fP\fR( 
\fB      unsigned short \fBctn\fR\fR, 
\fB      unsigned short \fBpn\fR\fR); 
.sp 1 
\fBchar \fBCT_data\fP\fR( 
\fB      unsigned short \fBctn\fR\fR, 
\fB      unsigned char * \fBdad\fR\fR, 
\fB      unsigned char * \fBsad\fR\fR, 
\fB      unsigned short \fBlenc\fR\fR, 
\fB      unsigned char * \fBcommand\fR\fR, 
\fB      unsigned short * \fBlenr\fR\fR, 
\fB      unsigned char * \fBresponse\fR\fR); 
.sp 1 
\fBchar \fBCT_close\fP\fR( 
\fB      unsigned short \fBctn\fR\fR); 
.fi 
.SH "DESCRIPTION" 
.PP 
This manual page describes the CT-API functions used to handle 
communication with integrated circuit cards and cardterminals. The  
CT-API functions are generic, and allows the use of memory cards (often  
referred as synchronous cards) and processor cards (often referred  
as asynchronous cards). 
 
.PP 
The CT-API implementation is dependent on the cardterminal that is  
used. For each cardterminal a CT-API library implementation  
should be provided, either by the manufacturer or by third parties.  
This manpage is not a complete specification of the CT-API, but  
a general overview of the most common functionalities. 
 
.PP 
\fBCT_init()\fP selects and initializes the  
interface with the cardterminal (f.i. serial port, USB bus, or any  
other type of connector). The CT_init() function must be called before  
communicating with the card or the cardterminal. A unique cardterminal  
number chosen by the caller is assigned by the CT-API library to the  
initialized connection. 
 
.IP "\fBctn\fR" 10 
Cardterminal number: identifies the connection with the 
cardterminal for further references. It's chosen by the caller 
of the function. 
 
.IP "\fBpn\fR" 10 
Port number: selects the physical interface. The port number 
assignment is dependent on the CT-API implementation. However most 
serial cardterminal implementations use the convention of assigning  
the value 0 to the first serial port, 1 to the second serial port and  
so on. See the file \fB/usr/include/ctapi.h\fP of 
your CT-API implementation for a list of allowed values. 
 
.PP 
\fBCT_data()\fP sends a command to the card or to  
the cardterminal and returns the response to the calling program. 
 
.IP "\fBctn\fR" 10 
Cardterminal number: the number specified in \fBCT_init() 
\fP call for this cardterminal. 
 
.IP "\fBdad\fR" 10 
Destination address. If the command is sent to the cardterminal,  
it must point to a variable containing the value CT (0x01). If the  
command is sent to a card, the destination address variable must be  
set to the number identifying the slot of the cardterminal where the  
card is inserted, ranging from ICC1 (0x00) and ICC2 (0x02) to ICC14  
(0x0E). 
 
.IP "" 10 
On the return of the response, the variable pointed by dad is set to  
the value HOST (0x02) or REMOTE_HOST (0x05). 
 
.IP "\fBsad\fR" 10 
Source address. When the function is called it is usually set to  
point to a variable with the value HOST (0x02), although the value 
REMOTE_HOST (0x05) may ocurr in some applications if the CT-API  
implementation supports remote calls. 
 
.IP "" 10 
On the return of the response, the variable pointed by sad is set to any  
of the values from CT (0x01), ICC1 (0x00), ICC2 (0x02) to ICC14 (0x0E). 
 
.IP "\fBlenc\fR" 10 
Command length in bytes. 
 
.IP "\fBcommand\fR" 10 
Pointer to the buffer that conveys the command to be sent 
to the card or cardterminal. 
 
.IP "\fBlenr\fR" 10 
Before function call, pointer to a variable containig the  
maximum buffer size in bytes for the response data. After  
successful function call it contains the size in bytes of the  
response data. 
 
.IP "\fBresponse\fR" 10 
Pointer to the response data. Memory buffer of at least the size 
in bytes specified in \fBlenr\fR must be allocated  
before calling to the function. 
 
.PP 
\fBCT_close()\fP terminates the communication  
with the cardterminal which has been assigned to a logical cardterminal 
number by the function \fBCT_init()\fP. The function  
shall be called for each initialized cardtreminal before the end of  
the program in order to free resources if necessary. 
 
.IP "\fBctn\fR" 10 
Cardterminal number: as specified in \fBCT_init() 
\fP call for this cardterminal. 
 
.SH "RETURN VALUE" 
.PP 
\fBCT_init(),\fP \fBCT_data(),\fP         and \fBCT_close()\fP functions return a value of type 
char. 
 
.IP "\fBOK\fR" 10 
Function call was successful. 
 
.IP "\fBERR_INVALID\fR" 10 
Invalid parameter or value. 
 
.IP "\fBERR_CT\fR" 10 
Cardterminal Error. The cardterminal is temporarily not 
accessible (busy with  other or internal processes). The problem can 
be solved by the application. 
 
.IP "\fBERR_TRANS\fR" 10 
Transmission Error. Transmission errors due to mechanical,  
electrical or protocol failures. Reset of the cardterminal is  
necessary. 
 
.IP "\fBERR_MEMORY\fR" 10 
Memory assignment error. A memory error occurred (f.i. the  
allocated buffer is too small for the returned data). 
 
.IP "\fBERR_HTSI\fR" 10 
Host Transport Service Interface error. Commonly returned if 
the error is produced by the software layer and not in the  
communication with the hardware. 
 
.SH "SEE ALSO" 
.PP 
ctbcs (3). 
.PP 
The \fBMKT\fP (Multifunktionale  
KartenTerminals) specifications, available for download from  
Teletrust website  
\fIhttp://www.teletrust.de (link to URL http://www.teletrust.de) \fR. 
In special the documents: 
.IP "Part 3:" 10 
CT-API. Cardterminal Applications Programming Interface. 
 
.IP "Part 4:" 10 
CT-BCS. Cardterminal Basic Command Set. 
 
.IP "Part 7:" 10 
IC cards with synchronous transmission Part3: Usage of  
interindustry Commands. 
 
.PP 
ISO/IEC 7816: \fBIdentification cards - Integrated circuit(s)  
cards with contacts\fP:  
 
.IP "Part 4:" 10 
Interindustry commands for interchange. 
 
.SH "AUTHOR" 
.PP 
This manual page was written by Carlos Prados  
<cprados@yahoo.com>. 
 
...\" created by instant / docbook-to-man, Sat 01 Dec 2001, 19:51 


NEWS
Unix driver for Towitoko smartcard readers
Copyright (C) 2000 2001 Carlos Prados <cprados@yahoo.com>

SUPPORT FOR PPS AND BAUDRATE GREATER THAN 9600 BPS

Starting with the release 2.0.5 it's possible to specify a PPS request to be
negotiated with the card. After completion of a successful PPS handshake,
the unit will be configured automatically to support the protocol, parameters
and baudrate agreed.

Baudrates of 9600 bps, 14400 bps, 19200 bps, 28800 bps, 38400 bps and 57600 bps
are supported by the driver.  

To access this functionality under CT-API, the application must issue a RESET CT
or REQUEST ICC and put the PPS request (at least PPSS and PPS0) in the data 
field of the CT-BCS command.

Under PCSC Lite the PPS request can be sent within a CT-BCS command, using the
function SCardControl().

Memory cards are accessed by default at 57600 bps if the unit supports this
baudrate, without having to modify the programs.

SUPPORT FOR CHIPDRIVE USB

Chipdrive USB is supported under Linux since release 2.0.5. You need 
a Linux kernel version 2.4.12 or greater with USB support compiled 
and pl2303 USB to serial converter driver enabled.

To use an USB Chipdrive with CT-API, specify port numbers starting
with 0x8000:
* 0x8000 for /dev/ttyUSB0, (first USB serial device)
* 0x8001 for /dev/ttyUSB1, (second USB serial device) and so on.

You can also create a symlink from /dev/ttyS[pn] to /dev/ttyUSB[m] and
access the reader at COM[pn+1], as with the serial model.

Under PCSC Lite just plug the reader and it will be automatically
detected and initialized by PCSC daemon. They name assigned by 
SCardListReaders will be "Towitoko Chipdrive USB 0 0".

SUPPORT FOR T=1 PROTOCOL

Smartcards with transmission protocol T=1 are supported since release
2.0.0.

When a card is reset, T=1 protocol is automatically selected if it is
offered by the card in its answer to reset. PTS protocol is performed
if  needed. Otherwise T=0 is used.

CHANGE IN THE NAME OF THE SHARED LIBRARY

Since release 2.0.0 the core driver and the interfaces are linked in a
single shared library, called libtowitoko.

In consequence the name of the library declared in PC/SC Lite
configuration file /etc/reader.conf must be changed to:

<path>/lib/libtowitoko.<extension>

Where <path> stands for the place where you installed the binary
distribution of the driver (like "/usr/local"), and <extension> is the
suffix of the name of shared libraries in you system (like "so").

A sample reader.conf file is included under doc/.

CHANGES IN THE INTERFACE WITH PC/SC LITE (release 2.0.0-pre5)

Version 2.0.0-pre5 comes with IFD handler 2.0. In contrast with
previous IFD handler, this module has been designed to support any
number of readers and slots per instance of the driver.

Also it permits the driver to return the ATR and its length when this
parameter is requested by PC/SC Lite, so it does not need to re-parse
the ATR. This allows memory cards, whose ATR does not follow ISO 7816
rules to be handled.

To simplify the /etc/reader.conf syntax, the channel number CHANNELID
now does not use the serial port IO and IRQ number but the COM number,
following MS-DOS numbering convention. That is 0x000001 for COM1
(/dev/ttyS0), 0x000002 for COM2 (/dev/ttyS1) and so on. Don't forget
to change this parameter in your /etc/reader.conf file if you upgrade
to 2.0.0-pre5.

Note that PC/SC Lite 0.7.1 or greater is needed to use IFD Handler
2.0.

CHANGES IN RELEASE 2.0.0-pre1

Since 1.X releases most of the code has been rewritten. The driver has
more than 8000 lines of code. Nevertheless now it's more modular and
easy to maintain. The new structure split the functionality between
modules that implement objects. Each object has the methods (C
functions) to handle a part of the communication, and the member
variables  (registers of C structures) to store status information
needed for this part of  the communication.

As result the new driver is able to handle any number of readers, with
any number of slots and smartcards at one time.

Object modules has been added to implement support for memory cards
and provide access through MCT v0.9 interindustry commands.

Interface device module has been improved to be able to dynamically
adapt its settings to support communication with inverse convention
cards and with Kartenzwerg readers.

The serial device code has been rewritten and improved. The new serial
object does PnP COM recognition of the reader on initialization. The
platform dependent code has been wrapped with #ifdef's to provide
Input/Output routines for Linux, Solaris and CygWin32 environments,
and make it extensible for other Operating Systems.


src/ct-api/ctapi.c
/*
    ctapi.c
    CT-API interface implementation for multiple card-terminals

    This file is part of the Unix driver for Towitoko smartcard readers
    Copyright (C) 1998 1999 2000 Carlos Prados <cprados@yahoo.com>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "defines.h"
#include "ctapi.h"
#include "ctbcs.h"
#include "ct_list.h"
#include "cardterminal.h"
#include "ct_slot.h"
#include <stdio.h>
#include <string.h>
#ifdef HAVE_PTHREAD_H
#include <pthread.h>
#endif

/* 
 * Not exported variables definition
 */

/* Linked list of card-terminals */
static CT_List *ct_list = NULL;

#ifdef HAVE_PTHREAD_H
static pthread_mutex_t ct_list_mutex = PTHREAD_MUTEX_INITIALIZER;
#endif

/*
 * Exported functions definition
 */

char
CT_init (unsigned short ctn, unsigned short pn)
{
  CardTerminal *ct;
  char ret;
  bool ct_list_empty;

#ifdef HAVE_PTHREAD_H
  pthread_mutex_lock (&ct_list_mutex);
#endif

  /* Check that ctn is not in use */
  if (CT_List_GetCardTerminal (ct_list, ctn) == NULL)
    {
      /* Create a new CardTerminal */
      ct = CardTerminal_New ();

      if (ct != NULL)
        {   
          /* Initialize CardTerminal */
          ret = CardTerminal_Init (ct, pn);

          /* Add CardTerminal to list */
          if (ret == OK)
            {  
              /* See if list is initialised */
              ct_list_empty = (ct_list == NULL);
              
              if (ct_list_empty)
                ct_list = CT_List_New ();

              /* Add the CardTerminal to the list */
              if (!CT_List_AddCardTerminal (ct_list, ct, ctn))
                {
                  CardTerminal_Close (ct);
                  CardTerminal_Delete (ct);
                  
                  if (ct_list_empty)
                    {
                      CT_List_Delete (ct_list);
                      ct_list = NULL;
                    }
                  
                  ret = ERR_MEMORY;
                }
            }
          
          else
            CardTerminal_Delete(ct);
        }
      else
        ret = ERR_MEMORY;
    }
  else
    ret = ERR_CT;

#ifdef HAVE_PTHREAD_H
  pthread_mutex_unlock (&ct_list_mutex);
#endif  

#ifdef DEBUG_CTAPI
  printf ("CTAPI: CT_init(ctn=%u, pn=%u)=%d\n", ctn, pn, ret);
#endif

  return ret;
}

char
CT_close (unsigned short ctn)
{
  CardTerminal *ct;
  char ret;

#ifdef HAVE_PTHREAD_H
  pthread_mutex_lock (&ct_list_mutex);
#endif

  /* Get the card-terminal */
  ct = CT_List_GetCardTerminal (ct_list, ctn);

  if (ct != NULL)
    {    
      /* Close CardTerminal */
      ret = CardTerminal_Close(ct);
  
      /* Remove card-terminal from list */
      CT_List_RemoveCardTerminal (ct_list, ctn);

      /* Delete the list if there are no more card-terminals */
      if (CT_List_GetNumberOfElements (ct_list) == 0)
        {
          CT_List_Delete (ct_list);
          ct_list = NULL;
        }
    }

  else
    ret = ERR_CT;

#ifdef HAVE_PTHREAD_H
  pthread_mutex_unlock (&ct_list_mutex);
#endif

#ifdef DEBUG_CTAPI
  printf ("CTAPI: CT_close(ctn=%d)=%u\n", ctn, ret);
#endif

  return ret;
}

char
CT_data (unsigned short ctn, unsigned char *dad, unsigned char *sad,
	 unsigned short lc, unsigned char *cmd, unsigned short *lr,
	 unsigned char *rsp)
{
  CardTerminal *ct;
  CT_Slot *slot;
  APDU_Cmd *apdu_cmd;
  APDU_Rsp *apdu_rsp = NULL;
  int remain;
  unsigned char aux;
  char ret;

#ifdef DEBUG_CTAPI
  int i;

  printf ("CTAPI: CT_data(ctn=%u, *dad=0x%02X, *sad=0x%02X, lc=%u, *cmd={", 
	  ctn, *dad, *sad, lc);

  for (i=0; i<lc; i++)
    printf ("%02X ", cmd[i]);

  printf ("}, *lr=%u, rsp=[])\n", *lr); 
#endif

#ifdef HAVE_PTHREAD_H
  pthread_mutex_lock (&ct_list_mutex);
#endif

  /* Get card-terminal */
  ct = CT_List_GetCardTerminal (ct_list, ctn);

#ifdef HAVE_PTHREAD_H
  pthread_mutex_unlock (&ct_list_mutex);
#endif

  if (ct != NULL)
    {
      /* Create a command APDU */
      apdu_cmd = APDU_Cmd_New (cmd, lc);

      if (apdu_cmd != NULL)
        {

#ifdef HAVE_PTHREAD_H
          pthread_mutex_lock (CardTerminal_GetMutex(ct));
#endif

          /* Command goes to the reader */
          if ((*dad) == 1)
            {
              /* CT-BCS command */
              ret = CardTerminal_Command (ct, apdu_cmd, &apdu_rsp);

              (*sad) = 1;
              (*dad) = (*sad);
            }

          /* Command goes to an ICC */
          else 
            {
              /* Get the slot */
              slot = CardTerminal_GetSlot(ct, ((*dad)==0)? 0: (*dad)-1);

              if (slot != NULL)
                {
                  /* ICC command */
                  ret = CT_Slot_Command (slot, apdu_cmd, &apdu_rsp);
        
                  if (CT_Slot_GetICCType (slot) != CT_SLOT_NULL)
                    {
                      aux = (*sad);
                      (*sad) = (*dad);
                      (*dad) = aux;
	            }
                  else
	            {
	              (*dad) = (*sad);
	              (*sad) = 1;
	            }
                }
      
              else
                {
                  /* Invalid DAD address */
                  (*dad) = (*sad);
                  (*sad) = 1;
                  apdu_rsp = NULL;

                  ret = ERR_INVALID;
                }
            }

#ifdef HAVE_PTHREAD_H
          pthread_mutex_unlock (CardTerminal_GetMutex(ct));
#endif

          if (apdu_rsp != NULL)
            {
              /* Copy APDU data to rsp */
              remain = MAX ((short)APDU_Rsp_RawLen(apdu_rsp) - (*lr),0);

              if (remain > 0)
	        ret = ERR_MEMORY;

              (*lr) = MIN ((*lr), (short)APDU_Rsp_RawLen (apdu_rsp));
         
              memcpy (rsp, APDU_Rsp_Raw (apdu_rsp) + remain, (*lr));

              /* Delete response APDU */
              APDU_Rsp_Delete (apdu_rsp);
            }

          else 
            (*lr) = 0;

          /* Delete command APDU */
          APDU_Cmd_Delete (apdu_cmd);
	}
      else
        ret = ERR_MEMORY;
    }     
  else
    ret = ERR_CT;
  
#ifdef DEBUG_CTAPI
  printf ("CTAPI: CT_data(ctn=%u, *dad=0x%02X, *sad=0x%02X, lc=%u, *cmd={}, *lr=%u, rsp={", 
	  ctn, *dad, *sad, lc, *lr);

  for (i=0; i<*lr; i++)
    printf ("%02X ", rsp[i]);

  printf ("})=%d\n", ret); 
#endif

  return ret;
}
